module/zfs/zthr.c

```diff
@@ -241,42 +270,68 @@ zthr_wakeup(zthr_t *t)
 }
 
 /*
- * Note: If the zthr is not running (e.g. has been cancelled
+ * Sends a cancel request to the zthr and blocks until the zthr is
+ * blocked. If the zthr is not running (e.g. has been cancelled
  * already), this is a no-op.
  */
-int
+void
 zthr_cancel(zthr_t *t)
 {
-	int rc = 0;
-
 	mutex_enter(&t->zthr_lock);
 
-	/* broadcast in case the zthr is sleeping */
-	cv_broadcast(&t->zthr_cv);
+	if (t->zthr_thread != NULL) {
+		/* broadcast in case the zthr is sleeping */
+		cv_broadcast(&t->zthr_cv);
 
-	t->zthr_cancel = B_TRUE;
-	while (t->zthr_thread != NULL)
-		cv_wait(&t->zthr_cv, &t->zthr_lock);
-	t->zthr_cancel = B_FALSE;
-	rc = t->zthr_rc;
-	mutex_exit(&t->zthr_lock);
+		/*
+		 * We raise the cancellation flag so the zthr gets notified.
+		 *
+		 * We also save the current cancellation stamp and compare it
+		 * to the value that it has next time we wake up. The point
+		 * of this is to ensure that we (e.g. this thread) don't hang
+		 * when the zthr gets cancelled from us, but then immediately
+		 * resumed from somewhere else (e.g. zthr_thread gets
+		 * repopulated) before we get to wake up and exit this function.
+		 */
+		uint64_t stamp = t->zthr_cancellation_stamp;
+		t->zthr_cancel = B_TRUE;
+		while (t->zthr_thread != NULL &&
+		    stamp == t->zthr_cancellation_stamp) {
+			cv_wait(&t->zthr_cv, &t->zthr_lock);
+		}
+	}
 
-	return (rc);
+	mutex_exit(&t->zthr_lock);
 }
 
+/*
+ * Sends a resume request to the zthr and blocks until the zthr is
+ * running again. If the zthr is already running, this is a no-op.
+ */
 void
 zthr_resume(zthr_t *t)
 {
-	ASSERT3P(t->zthr_thread, ==, NULL);
-
 	mutex_enter(&t->zthr_lock);
 
-	ASSERT3P(&t->zthr_checkfunc, !=, NULL);
-	ASSERT3P(&t->zthr_func, !=, NULL);
-	ASSERT(!t->zthr_cancel);
+	if (t->zthr_thread == NULL) {
+		ASSERT3P(&t->zthr_checkfunc, !=, NULL);
+		ASSERT3P(&t->zthr_func, !=, NULL);
+		ASSERT(!t->zthr_cancel);
 
-	t->zthr_thread = thread_create(NULL, 0, zthr_procedure, t,
-	    0, &p0, TS_RUN, minclsyspri);
+		t->zthr_thread = thread_create(NULL, 0, zthr_procedure, t,
+		    0, &p0, TS_RUN, minclsyspri);
+
+		/*
+		* Wait until the zthr has finished initializing.
+		* This ensures that the only time another thread
+		* can obtain the lock is when the zthr func is
+		* running or the zthr is asleep.
```

Comment:
How does zthr_initializing accomplish this?  It looks like it's just used when starting/resuming the thread to make sure we don't return until the thread has grabbed the lock.  So maybe it helps *this* thread not be able to call back into the zthr code before the thread has started running, but I think other threads could still call concurrently and see `zthr_initializing=TRUE`.